cmd/juju/cloud/add_test.go

```diff
@@ -187,240 +271,225 @@ func (s *addSuite) TestInteractive(c *gc.C) {
-    regions:
-      london:
-        endpoint: http://london/1.0
-  man:
-    type: manual
-    endpoint: 192.168.1.6
-`[1:])
+	c.Check(numCallsToWrite(), gc.Equals, 1)
 }
 
 func (s *addSuite) TestInteractiveVSphere(c *gc.C) {
-	s.createTestCloudData(c)
-	command := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(map[string]cloudfile.Cloud{}, nil)
+	vsphereCloud := cloudfile.Cloud{
+		Type:      "vsphere",
+		AuthTypes: []cloudfile.AuthType{"userpass", "access-key"},
+		Endpoint:  "192.168.1.6",
+		Regions: []cloudfile.Region{
+			{
+				Name:     "foo",
+				Endpoint: "192.168.1.6",
+			},
+			{
+				Name:     "bar",
+				Endpoint: "192.168.1.6",
+			},
+		},
+	}
+	fake.Call("ParseOneCloud", []byte("auth-types:\n- userpass\nendpoint: 192.168.1.6\nregions:\n  bar: {}\n  foo: {}\n")).Returns(vsphereCloud, nil)
+	vsphereMetadata := map[string]cloudfile.Cloud{"mvs": vsphereCloud}
+	numCallsToWrite := fake.Call("WritePersonalCloudMetadata", vsphereMetadata).Returns(nil)
+
+	command := cloud.NewAddCloudCommand(fake)
 	err := testing.InitCommand(command, nil)
 	c.Assert(err, jc.ErrorIsNil)
 
 	ctx := &cmd.Context{
-		Dir:    c.MkDir(),
 		Stdout: ioutil.Discard,
 		Stderr: ioutil.Discard,
-		Stdin: strings.NewReader(`
-vsphere
-mvs
-192.168.1.6
-foo
-y
-bar
-n
-`[1:])}
+		Stdin: strings.NewReader("" +
+			/* Select cloud type: */ "vsphere\n" +
+			/* Enter a name for the cloud: */ "mvs\n" +
+			/* Enter the controller's hostname or IP address: */ "192.168.1.6\n" +
+			/* Enter region name: */ "foo\n" +
+			/* Enter another region? (Y/n): */ "y\n" +
+			/* Enter region name: */ "bar\n" +
+			/* Enter another region? (Y/n): */ "n\n",
+		),
+	}
 
 	err = command.Run(ctx)
 	c.Check(err, jc.ErrorIsNil)
 
-	data, err := ioutil.ReadFile(osenv.JujuXDGDataHomePath("clouds.yaml"))
-
-	// Yes, the conversion copies the endpoint to each region, for some reason.
-	c.Assert(string(data), gc.Equals, `
-clouds:
-  homestack:
-    type: openstack
-    auth-types: [userpass, access-key]
-    endpoint: http://homestack
-    regions:
-      london:
-        endpoint: http://london/1.0
-  mvs:
-    type: vsphere
-    auth-types: [userpass]
-    endpoint: 192.168.1.6
-    regions:
-      bar:
-        endpoint: 192.168.1.6
-      foo:
-        endpoint: 192.168.1.6
-`[1:])
+	c.Check(numCallsToWrite(), gc.Equals, 1)
 }
 
 func (s *addSuite) TestInteractiveExistingNameOverride(c *gc.C) {
-	s.createTestCloudData(c)
-	command := cloud.NewAddCloudCommand()
+	fake := newFakeCloudQueryer()
+	fake.Call("PublicCloudMetadata", []string(nil)).Returns(map[string]cloudfile.Cloud{}, false, nil)
+	fake.Call("PersonalCloudMetadata").Returns(homestackMetadata(), nil)
+	manMetadata := map[string]cloudfile.Cloud{"homestack": manualCloud}
+	fake.Call("ParseOneCloud", []byte("endpoint: 192.168.1.6\n")).Returns(manualCloud, nil)
```

Comment:
It's not a gap in testing.  A change to ParseOneCloud would cause this test to fail. The way to fix this test would be to fix the code for add-cloud so it produces the correct input for ParseOneCloud.  Then the test would pass again, without having to update the test at all.  That's the ideal.  If you have to update the test every time something internal to the system changes, then the test knows too much about what it's testing.  Sometimes it's unavoidable, but in this case, it's not.

The test should say "whatever we output from add-cloud should be parseable by ParseOneCloud".  It doesn't actually care what we output so long as it parses into the right value.  The test shouldn't care if ParseOneCloud takes yaml or json, it should care that add-cloud produces -something- that produces the correct output from ParseOneCloud.